src/mito2/src/read/compat.rs

```diff
@@ -255,8 +271,21 @@ fn may_compat_fields(
         .iter()
         .map(|column_id| {
             if let Some(index) = source_field_index.get(column_id) {
+                // TODO(kould): more Modify case, e.g. column name
+                let mut modify_type = None;
+
+                // Safety: check on [`TableMeta::modify_columns`].
+                let expect_column = expect.column_by_id(*column_id).unwrap();
+                let actual_column = actual.column_by_id(*column_id).unwrap();
+
+                if expect_column != actual_column {
+                    modify_type = Some(expect_column.column_schema.data_type.clone())
+                }
```

Comment:
Compare the whole metadata here may introduce the same issue mentioned in https://github.com/GreptimeTeam/greptimedb/pull/3745#discussion_r1572049239

We can reuse the column type if `batch_fields()` and `projected_fields()` can return it. Then we don't need to call `column_by_id` to acquire the metadata again.